Add support of AES-GCM to PKCS11 TA (take 2)

etienne-lms opened the pull request:
This P-R supersedes #6467.

maroueneboubakri commented on ta/pkcs11/src/processing_symm.c:
```diff
@@ -263,6 +266,20 @@ allocate_tee_operation(struct pkcs11_session *session,
+	if (res == TEE_SUCCESS || params->id == PKCS11_CKM_AES_GCM) {
+		/*
+		 * Allocate a 2nd operation handler to save the operation state
+		 * on AES GCM one-short processing that queries the output
```
Typo: shot*

jforissier commented on ta/pkcs11/src/processing_aes.c:
```diff
@@ -14,6 +15,410 @@
+		if (!data_size) {
+			TEE_Free(data_ptr);
+			data_ptr = NULL;
+			DMSG("Is this expected from the Core API?");
```
Should not happen indeed. I believe `assert(data_size);` would be better here.

jforissier commented on ta/pkcs11/src/processing_symm.c:
```diff
@@ -963,6 +1024,43 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
+				out_size = ae_out_size;
+			break;
+		case PKCS11_FUNCTION_DECRYPT:
+			/* Now we're ready to releave data */
```
s/releave/release/

maroueneboubakri replied:
I believe should be reveal.

etienne-lms replied:
indeed, thanks

jforissier reviewed (approved):
- For "ta: pkcs11: pkcs11_ta.h: define PKCS11_CKM_AES_GCM":
  `Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`
- For "ta: pkcs11: implement AES GCM operations":
  With minor comments
  `Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

etienne-lms commented:
Comments addressed.
I've added a commit to the series: "ta: pkcs11: factorize second operation handle".

jforissier commented on ta/pkcs11/src/processing_aes.c:
```diff
@@ -332,12 +332,7 @@ enum pkcs11_rc tee_ae_decrypt_final(struct pkcs11_session *session,
-			data_ptr = NULL;
-			DMSG("Is this expected from the Core API?");
-		}
+		assert(!res && data_size);
```
Asserting `!res` may be a bit much, runtime errors can happen. However `data_size` should never be set to zero on output so `assert(data_size)` does make sense (but may not be that useful).

etienne-lms replied:
Ok, i'll fix that. Thanks.

jforissier commented:
For "ta: pkcs11: factorize second operation handle":

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

etienne-lms commented:
Review comment addressed.
@maroueneboubakri, if you're fine with this change, could you post a review tag?
@vesajaaskelainen, maybe would you like to have a look at this change?

maroueneboubakri commented:
`Reviewed-by: Marouene Boubakri <marouene.boubakri@nxp.com>`

etienne-lms commented:
I've squashed the fixup commits, applied the review tags and rebased the series.
Thanks for the reviews.

@jforissier, I've applied your A-b tag on commit "ta: pkcs11: implement AES GCM operations", despite I did not ask you if the fixup commit for https://github.com/OP-TEE/optee_os/pull/6683#pullrequestreview-1882811641 was fine. Please tell me if it's not ok.

etienne-lms commented:
I would need review tags on the optee_test counter part: https://github.com/OP-TEE/optee_test/pull/730.

jforissier commented:
@etienne-lms sure!
